src/edu/csus/ecs/pc2/ui/ShadowCompareRunsPane.java

```diff
@@ -303,10 +361,372 @@ public void actionPerformed(ActionEvent arg0) {
+     * 
+     * @return the count of selected runs (rows) in the runs table.
+     */
+    private int getCountOfSelectedRuns() {
+        if (resultsTable==null) {
+            return 0;
+        } else {
+            return resultsTable.getSelectedRowCount();
+        }
+    }
+
+    /**
+     * Processes each currently-selected row in the runs table by invoking the PC2 server to update the judgement for the submission
+     * listed in that row to match the judgement specified by the remote (Primary) CCS, unless either PC2 or the
+     * remote CCS shows the judgement as "Pending".
+     * 
+     * @return a count of the number of runs that were successfully updated.
+     * 
+     * 
+     */
+    private int updateSelectedRuns() {
+
+        //debug:
+//        System.out.println ("Updating the following " + getCountOfSelectedRuns() + " runs in PC2:");
+        
+        //get JTable (view) row indices
+        int [] viewRowIndices = resultsTable.getSelectedRows();
+        
+        //convert view indices to model indices
+        int [] modelRowIndices = new int[viewRowIndices.length];
+        for (int i=0; i<viewRowIndices.length; i++) {
+            modelRowIndices[i] = resultsTable.convertRowIndexToModel(viewRowIndices[i]);
+        }
+        
+        //build a map of the submissionIds of all the selected runs (rows) and the corresponding model row
+        HashMap<Integer, Integer> submissionIdMap = new HashMap<Integer,Integer>();
+        for (int modelRow : modelRowIndices) {
+            
+            //get the next submission id
+            //TODO: use an Enum for columnIds instead of hard-coded integers, which can change if the table is rearranged
+            Integer nextSubmissionId = (Integer) resultsTable.getModel().getValueAt(modelRow, 3); // 3=SubmissionID
+
+            //add the submissionID to the map of submissions to be updated, attaching its model row number
+            submissionIdMap.put(nextSubmissionId, modelRow);
+            
+            //debug:
+//            System.out.println ("   Submision ID: " + nextSubmissionId);
+            
+        }
+        
+        int numRunsUpdated = 0;
+        for (int submissionId : submissionIdMap.keySet()) {
+                        
+            //make sure the run is not "Pending" in either system
+            Integer modelRow = submissionIdMap.get(submissionId);
+            String pc2Judgement = (String) resultsTable.getModel().getValueAt(modelRow, 4);      // 4 = PC2 Judgment
+            String remoteJudgement = (String) resultsTable.getModel().getValueAt(modelRow, 5);   // 5=remote CCS judgement
+            
+            if (pc2Judgement.toLowerCase().contains("pending") || remoteJudgement.toLowerCase().contains("pending")) {
+                
+                JOptionPane.showMessageDialog(this, "Cannot update submission " + submissionId + " because it still has judgements Pending", 
```

Comment:
This message dialog is in a loop, consider if there are 100 run judgements
that cannot be updated the user will have to click ok 100 times.
If this is not likely, then no change is needed.   

Wanted to note that *in general* showing a message dialog in a loop is a bad
idea - but if the likelihood is low that there will be more than a couple
times this dialog appears - then it is ok.

Reply:
At this point in the code it is processing exactly (only) the runs which the user has selected.  The only way the "100 times" scenario could happen in this case is if the user selected 100 _pending_ runs and told the system to resolve them.  If they do that, I think they DESERVE to have to click OK 100 times -- it will teach them not to try to resolve Pending (unjudged) runs.